engine/server/sv_client.c

```diff
@@ -2119,6 +2119,780 @@ static qboolean SV_SendBuildInfo_f( sv_client_t *cl )
+	MSG_WriteByte( &cl->netchan.message, svc_stufftext );
+	MSG_WriteString( &cl->netchan.message, "ent_getvars_cb\n" ); // why do we need this?
+}
+
+/*
+===============
+SV_EntCreate_f
+
+Create new entity with specified name.
+===============
+*/
+static qboolean SV_EntCreate_f( sv_client_t *cl )
+{
+	edict_t	*ent = NULL;
+	int	i = 0;
+	string_t classname;
+
+	if( Cmd_Argc() < 2 )
+	{
+		SV_ClientPrintf( cl, "Use ent_create <classname> <key1> <value1> <key2> <value2> ...\n" );
+		return false;
+	}
+
+	classname = ALLOC_STRING( Cmd_Argv( 1 ) );
+
+	ent = SV_AllocPrivateData( 0, classname );
+
+	// Xash3D extension
+	if( !ent && svgame.physFuncs.SV_CreateEntity )
+	{
+		ent = SV_AllocEdict();
+		ent->v.classname = classname;
+		if( svgame.physFuncs.SV_CreateEntity( ent, (char*)STRING( classname ) ) == -1 )
+		{
+			if( ent && !ent->free )
+				SV_FreeEdict( ent );
+			ent = NULL;
+		}
+	}
+
+	// XashXT does not implement SV_CreateEntity, use saverestore export
+	if( !ent && svgame.physFuncs.pfnCreateEntitiesInRestoreList )
+	{
+		SAVERESTOREDATA data = { 0 };
+		ENTITYTABLE table = { 0 };
+		data.tableCount = 1;
+		data.pTable = &table;
+		table.classname = classname;
+		table.id = -1;
+		table.size = 1;
+		svgame.physFuncs.pfnCreateEntitiesInRestoreList( &data, 0, false );
+		ent = table.pent;
+	}
+
+	if( !ent )
+	{
+		SV_ClientPrintf( cl, "Invalid entity!\n" );
+		return false;
+	}
+
+	// choose default origin
+	ent->v.origin[2] = cl->edict->v.origin[2] + 25;
+	ent->v.origin[1] = cl->edict->v.origin[1] + 100 * sin( DEG2RAD( cl->edict->v.angles[1] ) );
+	ent->v.origin[0] = cl->edict->v.origin[0] + 100 * cos( DEG2RAD( cl->edict->v.angles[1] ) );
+
+	SV_LinkEdict( ent, false );
+
+	// apply keyvalues if supported
+	if( svgame.dllFuncs.pfnKeyValue )
+	{
+		for( i = 2; i < Cmd_Argc() - 1; i++ )
+		{
+			string keyname;
+			string value;
+			KeyValueData pkvd;
+
+			// allow split keyvalues to prespawn and postspawn
+			if( !Q_strcmp( Cmd_Argv( i ), "|" ) )
+				break;
+
+			Q_strncpy( keyname, Cmd_Argv( i++ ), sizeof( keyname ) - 1 );
+			Q_strncpy( value, Cmd_Argv( i ), sizeof( value ) - 1);
+			pkvd.fHandled = false;
+			pkvd.szClassName = (char*)STRING( ent->v.classname );
+			pkvd.szKeyName = keyname;
+			pkvd.szValue = value;
+			svgame.dllFuncs.pfnKeyValue( ent, &pkvd );
+
+			if( pkvd.fHandled )
+				SV_ClientPrintf( cl, "value \"%s\" set to \"%s\"!\n", pkvd.szKeyName, pkvd.szValue );
+		}
+	}
+
+	// set default targetname
+	if( !ent->v.targetname )
+	{
+		string newname, clientname;
+		int j;
+
+		for( j = 0; j < sizeof( cl->name ); j++ )
+		{
+			char c = Q_tolower( cl->name[j] );
+			if( c < 'a' || c > 'z' )
+				c = '_';
+			if( !cl->name[j] )
+			{
+				clientname[j] = 0;
+				break;
+			}
+			clientname[j] = c;
+		}
+
+		// generate name based on nick name and index
+		Q_snprintf( newname, sizeof( newname ) - 1, "%s_%i_e%i", clientname, cl->userid, NUM_FOR_EDICT( ent ));
```

Comment:
> sizeof( newname ) - 1

It's incorrect here too because Q_snprintf implemented as `snprintf` from C99.
`_snprintf` from msvcrt has another behavior:
https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/snprintf-snprintf-snprintf-l-snwprintf-snwprintf-l?view=msvc-170
> The snprintf function and the _snprintf family of functions format and store count or fewer characters in buffer. The snprintf function always stores a terminating null character, truncating the output if necessary. The _snprintf family of functions only appends a terminating null character if the formatted string length is strictly less than count characters. 

So there no reason to reduce size limit too.